File: core/src/main/java/org/mifos/mobilewallet/core/domain/usecase/CreateClient.java
package org.mifos.mobilewallet.core.domain.usecase;

import org.mifos.mobilewallet.core.base.UseCase;
import org.mifos.mobilewallet.core.data.fineract.repository.FineractRepository;
import org.mifos.mobilewallet.core.domain.model.NewClient;

import javax.inject.Inject;

/**
 * Created by naman on 20/8/17.
 */

public class CreateClient extends UseCase<CreateClient.RequestValues, CreateClient.ResponseValue> {

    private final FineractRepository apiRepository;

    @Inject
    public CreateClient(FineractRepository apiRepository) {
        this.apiRepository = apiRepository;
    }

    @Override
    protected void executeUseCase(RequestValues requestValues) {

    }

    public static final class RequestValues implements UseCase.RequestValues {

        private final NewClient client;

        public RequestValues(NewClient client) {
            this.client = client;
        }
    }

    public static final class ResponseValue implements UseCase.ResponseValue {


        public ResponseValue() {
        }

    }

}

File: core/data/src/androidMain/java/org/mifospay/core/data/domain/usecase/client/CreateClient.kt
/*
 * Copyright 2024 Mifos Initiative
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * See https://github.com/openMF/mobile-wallet/blob/master/LICENSE.md
 */
package org.mifospay.core.data.domain.usecase.client

import com.mifospay.core.model.domain.client.NewClient
import org.mifospay.core.data.base.UseCase
import org.mifospay.core.data.fineract.repository.FineractRepository
import org.mifospay.core.data.util.ErrorJsonMessageHelper.getUserMessage
import retrofit2.HttpException
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers

class CreateClient(
    private val apiRepository: FineractRepository,
) : UseCase<CreateClient.RequestValues, CreateClient.ResponseValue>() {

    data class RequestValues(val client: NewClient) : UseCase.RequestValues

    data class ResponseValue(val clientId: Int) : UseCase.ResponseValue

    override fun executeUseCase(requestValues: RequestValues) {
        apiRepository.createClient(requestValues.client)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(
                object : Subscriber<ResponseValue>() {
                    override fun onCompleted() {}
                    override fun onError(e: Throwable) {
                        var message: String
                        try {
                            message =
                                (e as HttpException).response()?.errorBody()?.string().toString()
                            message = getUserMessage(message)
                        } catch (e1: Exception) {
                            message = e1.message.toString()
                        }
                        useCaseCallback.onError(message)
                    }

                    override fun onNext(genericResponse: ResponseValue) {
                        useCaseCallback.onSuccess(genericResponse)
                    }
                },
            )
    }
}
